### ISSUE TITLE
Tpetra host lambda CUDA build fail

### ISSUE THREAD

**Issue by ibaned** [trilinos/Trilinos#1717]
Nightly builds of Albany are failing with the following:

```
Trilinos/packages/kokkos/core/src/Cuda/Kokkos_CudaExec.hpp(260): error: The closure type for a lambda ("lambda [](size_t)->void") cannot be used in the template argument type of a __global__ function template instantiation, unless the lambda is defined within a __device__ or __global__ function, or the lambda is a 'extended lambda' and the flag --expt-extended-lambda is specified
          detected during:
            instantiation of "Kokkos::Impl::cuda_parallel_launch_local_memory" based on template arguments void, Kokkos::RangePolicy>, Kokkos::CudaUVMSpace::execution_space>, 1024U, 1U> 
(260): here
            instantiation of "Kokkos::Impl::CudaParallelLaunch::CudaParallelLaunch(const DriverType &, const dim3 &, const dim3 &, int, cudaStream_t) [with DriverType=Kokkos::Impl::ParallelForvoid, Kokkos::RangePolicy>, Kokkos::CudaUVMSpace::execution_space>, LaunchBounds=Kokkos::LaunchBounds<1024U, 1U>]" 
/.../repos/Trilinos/packages/kokkos/core/src/Cuda/Kokkos_Cuda_Parallel.hpp(370): here
            instantiation of "void Kokkos::Impl::ParallelFor, Kokkos::Cuda>::execute() const [with FunctorType=lambda [](size_t)->void, Traits=>]" 
/.../repos/Trilinos/packages/kokkos/core/src/Kokkos_Parallel.hpp(191): here
            instantiation of "void Kokkos::parallel_for(const ExecPolicy &, const FunctorType &, const std::__cxx11::string &, Kokkos::Impl::enable_if<, void>::type *) [with ExecPolicy=Kokkos::RangePolicy>, FunctorType=lambda [](size_t)->void]" 
/.../repos/Trilinos/packages/tpetra/core/src/Tpetra_Details_unpackCrsMatrixAndCombine_def.hpp(952): here
            instantiation of "void Tpetra::Details::unpackAndCombineIntoCrsArraysImpl(const LocalMatrix &, const LocalMap &, const Tpetra::Details::PackTraits::input_array_type &, const Tpetra::Details::PackTraits::input_buffer_type &, const Tpetra::Details::PackTraits::input_array_type &, const Tpetra::Details::PackTraits::input_array_type &, const Tpetra::Details::PackTraits::input_array_type &, const Tpetra::Details::PackTraits::output_array_type &, const Tpetra::Details::PackTraits::output_array_type &, const Tpetra::Details::PackTraits::output_array_type &, const Tpetra::Details::PackTraits::input_array_type &, const Tpetra::Details::PackTraits::output_array_type &, size_t, size_t, size_t, int, size_t) [with LocalMatrix=KokkosSparse::CrsMatrix, LocalMap=Tpetra::Details::LocalMap>]" 
/.../repos/Trilinos/packages/tpetra/core/src/Tpetra_Details_unpackCrsMatrixAndCombine_def.hpp(1418): here
            instantiation of "void Tpetra::Details::unpackAndCombineIntoCrsArrays(const Tpetra::CrsMatrix &, const Teuchos::ArrayView &, const Teuchos::ArrayView &, const Teuchos::ArrayView &, size_t, Tpetra::Distributor &, Tpetra::CombineMode, size_t, const Teuchos::ArrayView &, const Teuchos::ArrayView &, size_t, size_t, int, const Teuchos::ArrayView &, const Teuchos::ArrayView &, const Teuchos::ArrayView::impl_scalar_type> &, const Teuchos::ArrayView &, Teuchos::Array &) [with Scalar=double, LocalOrdinal=int, GlobalOrdinal=int, Node=Tpetra::Kokkos_Compat_KokkosCudaWrapperNode]" 
/.../build/TriBuild/packages/tpetra/core/src/Tpetra_Details_unpackCrsMatrixAndCombine_DOUBLE_INT_INT_CUDA.cpp(78): here
```

Without looking at the code, it sounds to me like a non-Kokkos lambda was given to a Kokkos parallel construct.

@trilinos/tpetra

**Comment by tjfulle** [comment]
@ibaned, there are not any non-Kokkos lambda in the files in question, but we have been having problems with correct lambda compile time options (see #1682).  Can you add `--expt-extended-lambda` to the command line to see if it resolves the problem.

**Comment by ibaned** [comment]
You're right, the code is fairly benign:
```
  // Zero the rowptr
  parallel_for(range_policy(0, N+1),
    KOKKOS_LAMBDA(const size_t i) {
      tgt_rowptr(i) = 0;
    }
  );
```
I think what we're missing is `-DKokkos_ENABLE_Cuda_Lambda=ON`.
I also think this issue is just part of #1682 in general, and resolving #1682 is becoming critical as Tpetra starts to use lambdas.

**Comment by tjfulle** [comment]
@ibaned , ya, I need to shoot Ross some diagnostics of failing configures/builds so he can make necessary changes to TriBITS

**Comment by mhoemmen** [comment]
I can fix #1682; it's pretty straightforward.

**Comment by mhoemmen** [comment]
@ibaned I just pushed a fix for #1682 to develop.

Here is my reasoning behind allowing `KOKKOS_LAMBDA` in Tpetra:

  1. CUDA 7.5 and 8.0 both allow it
  2. Kokkos used to allow it by default, without any CMake options
  3. It saves time and makes code easier to read
  4. It is the future (support for `KOKKOS_LAMBDA` exists now and will only improve over time)

**Comment by ibaned** [comment]
@mhoemmen I totally agree with the use of lambdas, and just want to make it easy for this transition to happen. I'll push on the Kokkos side to have this Cuda_Lambda flag on by default.
